KAFKA-5111: Improve internal Task APIs

mjsax opened the pull request:
Refactors Task with proper interface methods `init()`, `resume()`, `commit()`, `suspend()`, and `close()`. All other methods for task handling are internal now. This allows to simplify `StreamThread` code, avoid code duplication and allows for easier reasoning of control flow.

mjsax committed:
Code Cleanup

mjsax committed:
KAFKA-5111: Improve internal Task APIs

mjsax commented:
Call for review @enothereska @dguy @guozhangwang

I did talk about this with @guozhangwang and he basically agrees that this refactoring make sense. It also allows to simplify the EoS code (for example, I could remove `closeProducer()`) and exception handling. Note, that this is 3 commits, while the second commit contains the actual refactoring. The other two are code cleanups only.

Please review extremely critically, as this is on the hot code path and we must get this correct to not make a step backwards. Nevertheless, I think the refactoring is a good idea to keep the code maintainable long term.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -110,15 +110,12 @@ public final ThreadCache cache() {
         return cache;
     }
 
+    public abstract void init();
```
Wondering if we should create an actual interface for this?

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -63,24 +63,24 @@
+        final Sensor taskCommitTimeSensor;
+
+
+        public TaskMetrics(final StreamsMetrics metrics) {
```
nit: remove `public`

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -190,10 +145,52 @@ public String toString(final String indent) {
         stateMgr.flush();
     }
+
+    void initializeStateStores() {
```
This can still be `protected` function?

mjsax replied:
"package private" is stronger than `protected` and there is no reason to make it more open than required. But I am open to make it `protected`, too. Just don't see why we would need it?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -190,10 +145,52 @@ public String toString(final String indent) {
+     * @throws ProcessorStateException if there is an error while closing the state manager
+     * @param writeCheckpoint boolean indicating if a checkpoint file should be written
+     */
+    void closeStateManager(final boolean writeCheckpoint) {
```
`protected`?

mjsax replied:
As above.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -281,6 +281,7 @@ public void close(Map<TopicPartition, Long> ackedOffsets) throws IOException {
     // write the checkpoint
     @Override
     public void checkpoint(final Map<TopicPartition, Long> ackedOffsets) {
+        log.trace("{} Writing checkpoint", logPrefix);
```
nit: including the acked offsets to checkpoint as well.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -97,31 +96,27 @@ public void commit() {
+        try {
+            commit();
+            closeStateManager(true);
+        } catch (final RuntimeException e) {
```
Since we are already capturing IOException internally, what other types of RTE could be thrown here?

mjsax replied:
Maybe there are none -- but isn't a general guard a good idea?

guozhangwang replied:
I'm just afraid that capturing any RTE that we have not thought about and re-close the state managers may hide some issues or even subsequently trigger some other issues.

guozhangwang replied:
EDIT: just realizing that we are re-throwing the exception anyways after re-closing the state managers. So this should be fine.

ijuma replied:
Hmm, this pattern is a bit weird. As @guozhangwang initially said, it's as likely to introduce issues than fix issues. `closeStateManager` should just do the right thing. If we need to change the `catch` blocks, we should do it there.

ijuma replied:
Or is your concern about `commit` throwing an exception? If that is the case, then the `try`/`catch` should just be around that method IMO.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java:
```diff
@@ -544,24 +539,19 @@ public void shouldCloseAllProcessorNodesWhenExceptionsRaised() throws Exception
     }
 
-    @Test(expected = IllegalStateException.class)
-    public void shouldThrowWhenClosingProducerForNonEoS() {
```
Is this deletion intentional?

mjsax replied:
Yes. `task.closeProducer()` get's removed by this PR.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1138,7 +1129,7 @@ protected StreamTask createStreamTask(final TaskId id, final Collection<TopicPar
 
     @Test
-    public void shouldNotViolateAtLeastOnceWhenExceptionOccursDuringCloseTopologyWhenSuspendingState() throws Exception {
+    public void shouldNotViolateAtLeastOnceWhenExceptionOccursDuringSuspendWhenSuspendingState() throws Exception {
```
nit: could this be further simplified to 

```
shouldNotViolateAtLeastOnceWhenExceptionOccursDuringTaskSuspension
```

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -190,10 +145,52 @@ public String toString(final String indent) {
+        return Collections.emptyMap();
+    }
+
+    protected void initializeOffsetLimits() {
```
This has stayed the same as before?

mjsax replied:
Yes. Just moved some methods in the third commit to "cluster" them a little bit in the code -- right now it's all scattered around the whole file and I have always are hard time to keep an overview (maybe it's just me...)

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -190,10 +145,52 @@ public String toString(final String indent) {
         stateMgr.flush();
     }
+
+    void initializeStateStores() {
```
Not sure what has changed here.

mjsax replied:
Just moved.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -190,10 +145,52 @@ public String toString(final String indent) {
+     * @throws ProcessorStateException if there is an error while closing the state manager
+     * @param writeCheckpoint boolean indicating if a checkpoint file should be written
+     */
+    void closeStateManager(final boolean writeCheckpoint) {
```
Not sure what has changed here.

mjsax replied:
Just moved.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -142,39 +142,31 @@ public StreamTask(final TaskId id,
-    public int numBuffered() {
-        return partitionGroup.numBuffered();
+    @Override
+    public void resume() {
```
What does resume mean? It looks like it's initing?

mjsax replied:
Yes it does. It's just to get a clean API -- tasks are "resumed" when they are re-assigned (ie, reused after the got suspended). I though it's helpful to have an extra method for this.

guozhangwang replied:
I thought that 

1. `init` include both initializing the topology (i.e. processorNode.init) as well as the state stores; 
2. and `suspend` does not close state stores, only flush them.
3. and `resume` only re-init the topology but not state stores.

cc @mjsax ?

mjsax replied:
Ack.
`suspend` already works like this.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -213,92 +201,93 @@ public boolean process() {
+        log.trace("{} Committing", logPrefix);
+        metrics.metrics.measureLatencyNs(
+            time,
+            new Runnable() {
```
I don't immediately see why this cleanup is needed here.

mjsax replied:
The class is only used once -- and this change keep code local -- when reading code before, it was unclear what's actually happening without stepping into the class that is on a completely different place in the file. Now you can read it at one glance.

I guess it's a personal thing and I revert if you don't like it.

enothereska commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -140,54 +138,228 @@ public boolean isValidTransition(final State newState) {
+     * This class extends {@link StreamsMetricsImpl(Metrics, String, String, Map)} and
+     * overrides one of its functions for efficiency
+     */
+    private class StreamsMetricsThreadImpl extends StreamsMetricsImpl {
```
I don't know what has changed here.

mjsax replied:
Just moved.

enothereska reviewed (commented):
I'm not sure why this cleanup blocks EoS. A lot of code has moved around, perhaps in a different style.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -71,6 +71,7 @@
 
     private static final Logger log = LoggerFactory.getLogger(StreamThread.class);
     private static final AtomicInteger STREAM_THREAD_ID_SEQUENCE = new AtomicInteger(1);
+
```
@mjsax This file has been a bit hard to review right now, with all the actual changes mixed with code style cleanups and function move-arounds. Could you only maintain the actual change for this PR and with the function move-arounds / cleanups in another PR?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -142,39 +142,31 @@ public StreamTask(final TaskId id,
         initializeStateStores();
         stateMgr.registerGlobalStateStores(topology.globalStateStores());
-        initTopology();
+        init();
```
nit: could we move the `initializeStateStores` as well as global state store registration inside `init` functions for both streamtask and standby task, since we are closing them inside `close()` as well?

mjsax replied:
Ack.

mjsax commented:
I will remove the last commit and extract it into an own PR, too.

mjsax committed:
Github comments

mjsax commented:
Updated this. (removed one cleanup commit)

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -193,7 +190,7 @@ public String toString(final String indent) {
      * Flush all state stores owned by this task
      */
-    public void flushState() {
+    protected void flushState() {
```
Could this be `no modifier`?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java:
```diff
@@ -300,23 +308,25 @@ public void checkpoint(final Map<TopicPartition, Long> ackedOffsets) {
 
     void registerGlobalStateStores(final List<StateStore> stateStores) {
-        for (StateStore stateStore : stateStores) {
+        log.info("{} Register global stores", logPrefix);
```
nit: print store names?

mjsax replied:
I had the same though be did not to it, because it will be the same log message for all processors... quite repetitive. Can add it though.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -49,45 +49,52 @@
     }
 
-    public Collection<TopicPartition> changeLogPartitions() {
+    Collection<TopicPartition> changeLogPartitions() {
```
This function seems only used in unit test now. In that case could we just call `checkpointedOffsets().keySet()`?

guozhangwang replied:
One more thing I realized while reviewing this PR: `updateStandByTaskMaps` is called in two different places but that seems duplicated: first in the middle of `addStandbyTasks`, then at the end of `addStandbyTasks` by `new StandbyTaskCreator(checkpointedOffsets).retryWithBackoff(newStandbyTasks, start);` which then calls this function. 

Would like you to double check if I'm right, if yes we can file a simple PR after this to remove one of its caller.

mjsax replied:
It's not good code but also not a bug atm. It would be sufficient to call `updateStandByTaskMaps` only for tasks we do resume -- if we don't have a task to resume, `updateStandByTaskMaps` will not so anything as `task` parameter is `null`. I guess I can just make this update within this PR -- why do you think we need an extra PR?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -49,45 +49,52 @@
     }
 
+    @Override
+    public void init() {}
```
Hmm, if standbyTask does not ever call `init()` then why do we want it to be an abstract class? I felt that we can either just remove abstract `init` and just let `StreamTask` has its own `init` function that is also called in `resume`; or if we do want to have a `init` abstract logic, we can also move

```
log.info("standby-task [{}] Initializing state stores", id());
initializeStateStores();
checkpointedOffsets = Collections.unmodifiableMap(stateMgr.checkpointed());
processorContext.initialized();
```

At let constructor call this `init`.

Personally I like the first option since we are already overriding the constructors anyways which always covers the `init` part, and just the stream task needs a reusable `init` part that is to be used in `resume`. But it's your call.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyTask.java:
```diff
@@ -49,45 +49,52 @@
+    public void init() {}
+
+    @Override
+    public void resume() {}
```
Maybe we should re-`initializeOffsetLimits` here as its committed offsets could have been updated?

mjsax replied:
How could the offsets have changed? When we enter rebalance, we commit and keep the offset in the suspended standby task -- if we get the standby task reassigned its up-to-date -- otherwise, we drop it anyway.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -334,18 +342,33 @@ public void needCommit() {
+     */
     @Override
-    public void initTopology() {
+    public void init() {
```
Hmm, seems `init()` is also only called once in the constructor of `StreamTask`, and only `initTopology` is reused in `resume`; in that case I'd more prefer to just remove this abstract function and inline that in the constructor (see my comment above).

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -185,20 +165,17 @@ public int numBuffered() {
 
         // if there is no record to process, return immediately
         if (record == null) {
-            requiresPoll = true;
```
Nice catch on this not-used-anymore :P

mjsax replied:
Intellij warnings FTW! Can just recommend to turn them on :)

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -240,33 +212,30 @@ private void updateProcessorContext(final ProcessorRecordContext recordContext,
      */
-    public boolean maybePunctuate() {
-        long timestamp = partitionGroup.timestamp();
+    boolean maybePunctuate() {
```
nit: add a newline between line 214 and 215?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -275,30 +244,69 @@ public void punctuate(ProcessorNode node, long timestamp) {
+    public void suspend() {
+        log.info("{} Suspending task", logPrefix);
+        closeTopology();
+        log.debug("{} Committing offsets", logPrefix);
```
We can remove this log as it duplicates with line 263 above.

mjsax replied:
The log level in 263 is `trace` -- but we can remove, as we get `info` log level in 288.

guozhangwang replied:
I think the `trace` on 263 is good enough, covering line 290. For line 288 it should be kept as is.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:
```diff
@@ -436,19 +468,20 @@ public String toString() {
 
     @Override
-    public void flushState() {
+    protected void flushState() {
```
Since now stream thread does not directly call this function anymore, could it be `no modifier`?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -110,24 +110,21 @@ public final ThreadCache cache() {
     }
 
+    public abstract void init();
+    public abstract void resume();
```
`init` and `resume` are not used in standby tasks at all, so maybe we can just remove them from the abstract tasks (see my comments below)?

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1069,23 +996,24 @@ private StandbyTask createStandbyTask(TaskId id, Collection<TopicPartition> part
                 log.debug("{} recycling old standby task {}", logPrefix, taskId);
                 suspendedStandbyTasks.remove(taskId);
-                task.initTopology();
+                task.init();
```
For standby tasks `init` is no-op, so we can just remove it? PS: logically we should call `resume` not `init` here, but the former is also no-op anyways..

guozhangwang reviewed (commented):
More comments left. Otherwise lgtm.

mjsax committed:
Guozhang's comments

mjsax commented:
Updated this.

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -111,23 +111,18 @@ public final ThreadCache cache() {
 
     public abstract void commit();
-
+    public abstract void suspend();
```
I think we should introduce an interface, `Task`? Doesn't necessarily need to be done in this PR, though

dguy commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1196,23 +1102,23 @@ public String toString() {
      */
     public String toString(final String indent) {
-        final StringBuilder sb = new StringBuilder(indent + "StreamsThread appId: " + this.applicationId + "\n");
+        final StringBuilder sb = new StringBuilder(indent + "StreamsThread appId: " + applicationId + "\n");
```
nit: I know this was here before, but why use a `StringBuilder` and then use `+` ?

dguy reviewed (approved):
LGTM

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -947,7 +947,7 @@ private void addStreamTasks(final Collection<TopicPartition> assignment, final l
                     final StreamTask task = findMatchingSuspendedTask(taskId, partitions);
                     if (task != null) {
-                        log.debug("{} recycling old task {}", logPrefix, taskId);
+                        log.debug("{} resuming old task {}", logPrefix, taskId);
```
nit: this is not introduced in this PR but, other places capitalize the first letter after log prefix.

mjsax committed:
Github comments and offline discussion
 - added resume() to AbstractTask
 - imporved StateStoreManager.close() to be more robust

mjsax commented:
Updated this. Will so a follow up PR to address some more logging issues and other minor comment (like StringBuilder).

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

mjsax commented:
Retest this please.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/AbstractTask.java:
```diff
@@ -120,11 +121,7 @@ public final ThreadCache cache() {
-        } catch (final IOException e) {
-            throw new ProcessorStateException("Error while closing the state manager", e);
-        }
+        stateMgr.close(writeCheckpoint ? recordCollectorOffsets() : null);
```
`void closeStateManager(final boolean writeCheckpoint) throws ProcessorStateException `?

mjsax replied:
Will add -- going to do a follow up cleanup PR anyway.

guozhangwang reviewed (approved):
Merged to trunk. A minor comment that can be incorporated in the follow-up PR.
